UnityProject/Assets/Scripts/Generators/OverlayGenerator.cs

```diff
@@ -0,0 +1,56 @@
+using UnityEngine;
+using System.Collections;
+
+public class OverlayGenerator
+{
+    private GameObject terrainFolder;
+    private GameObject gridGameObject;
+    private static float TERRAIN_GRID_OFFSET = 0.01f;
+
+    public OverlayGenerator(GameObject terrainFolder = null, GameObject gridGameObject = null)
+    {
+        this.terrainFolder = terrainFolder;
+        this.gridGameObject = gridGameObject;
+    }
+
+    public GameObject GenerateGridForTerrain(TerrainDTO terrain)
+    {
+        if (!gridGameObject)
+        {
+            gridGameObject = GetGridGameObject();
+        }
+        Vector3 newScale = new Vector3(TerrainGenerator.TerrainScalingFactor * terrain.width,
+                                       TerrainGenerator.TerrainScalingFactor,
+                                       TerrainGenerator.TerrainScalingFactor * terrain.height);
+        if (gridGameObject.transform.localScale != newScale)
+        {
+            gridGameObject.transform.localScale = newScale;
+        }
+        UpdateGridMaterial(terrain);
+        return gridGameObject;
+    }
+
+    private void UpdateGridMaterial(TerrainDTO terrain)
+    {
+        Material material = gridGameObject.GetComponent<Renderer>().sharedMaterial;
+        material.mainTextureScale = new Vector2(terrain.width, terrain.height);
+    }
+
+    private GameObject GetGridGameObject()
+    {
+        GameObject grid = GameObject.FindWithTag("Grid");
+        if (grid == null)

```

Comment:
We don't necessarily need to do this, but I would change the `==` to `is` as we can always overwrite the equality operator in the future and then we can get some nasty results here

Reply:
C# 7 or higher is needed for this, our project is  in C# 4 so I don't think I can do this